File: 数据结构/UnionFind.cpp
class UnionFind {
    vector<int> fa;
    vector<int> siz;  // 集合大小
public:
    int cc;  // 连通块个数
    UnionFind(int n) : fa(n), siz(n, 1), cc(n) { ranges::iota(fa, 0); }
    int get(int x) {
        if (fa[x] != x) fa[x] = get(fa[x]);
        return fa[x];
    }
    bool is_same(int x, int y) { return get(x) == get(y); }
    bool merge(int from, int to) {
        int x = get(from), y = get(to);
        if (x == y) return false;
        fa[x] = y;
        siz[y] += siz[x];
        cc--;
        return true;
    }
    int get_size(int x) {  // 查询x所在集合大小
        return siz[get(x)];
    }
};

File: Python/数据结构/UnionFind.py
class UnionFind:
    def __init__(self, n: int) -> None:
        self.fa = list(range(n))
        self.siz = [1] * n  # 集合大小
        self.cc = n  # 连通块个数

    def get(self, x: int) -> int:
        root = x
        while self.fa[root] != root:
            root = self.fa[root]
        while self.fa[x] != x:
            fa = self.fa[x]
            self.fa[x] = root
            x = fa
        return root

    def is_same(self, x: int, y: int) -> bool:
        return self.get(x) == self.get(y)

    def merge(self, from_: int, to: int) -> bool:
        x, y = self.get(from_), self.get(to)
        if x == y:
            return False
        self.fa[x] = y
        self.siz[y] += self.siz[x]
        self.cc -= 1
        return True

    # 查询x所在集合大小
    def get_size(self, x: int) -> int:
        return self.siz[self.get(x)]
